libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -122,78 +152,109 @@ impl Replacements {
+        self.map.insert(hash(k.as_ref()), v)
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::{parse_expression, parse_expression_pattern, scan};
+
+    fn parse_rule(prog: &str) -> ExprPat {
+        let (expr, _) = parse_expression_pattern(scan(prog));
+        expr
+    }
+
+    fn parse_expr(prog: &str) -> Expr {
+        match parse_expression(scan(prog)) {
+            (Stmt::Expr(expr), _) => expr,
+            _ => unreachable!(),
+        }
+    }
 
     mod replacements {
         use super::*;
 
         #[test]
         fn try_merge() {
-            let a = ExprPat::VarPat("a".into());
-            let b = ExprPat::VarPat("b".into());
-            let c = ExprPat::VarPat("c".into());
+            let a = Rc::new(ExprPat::VarPat("a".into()));
+            let b = Rc::new(ExprPat::VarPat("b".into()));
+            let c = Rc::new(ExprPat::VarPat("c".into()));
 
             let mut left = Replacements::default();
-            left.insert(a.clone(), Expr::Const(1.));
-            left.insert(b.clone(), Expr::Const(2.));
+            left.insert(&a, Expr::Const(1.).into());
+            left.insert(&b, Expr::Const(2.).into());
 
             let mut right = Replacements::default();
-            right.insert(b.clone(), Expr::Const(2.));
-            right.insert(c.clone(), Expr::Const(3.));
+            right.insert(&b, Expr::Const(2.).into());
+            right.insert(&c, Expr::Const(3.).into());
 
             let merged = Replacements::try_merge(left, right).unwrap();
             assert_eq!(merged.map.len(), 3);
-            assert_eq!(merged.map.get(&a).unwrap().to_string(), "1");
-            assert_eq!(merged.map.get(&b).unwrap().to_string(), "2");
-            assert_eq!(merged.map.get(&c).unwrap().to_string(), "3");
+            assert_eq!(merged.map.get(&hash(&a)).unwrap().to_string(), "1");
+            assert_eq!(merged.map.get(&hash(&b)).unwrap().to_string(), "2");
+            assert_eq!(merged.map.get(&hash(&c)).unwrap().to_string(), "3");
         }
 
         #[test]
         fn try_merge_overlapping_non_matching() {
-            let a = ExprPat::VarPat("a".into());
+            let a = Rc::new(ExprPat::VarPat("a".into()));
 
             let mut left = Replacements::default();
-            left.insert(a.clone(), Expr::Const(1.));
+            left.insert(&a, Expr::Const(1.).into());
 
             let mut right = Replacements::default();
-            right.insert(a, Expr::Const(2.));
+            right.insert(&a, Expr::Const(2.).into());
 
             let merged = Replacements::try_merge(left, right);
             assert!(merged.is_none());
         }
+
+        #[test]
+        fn transform_common_subexpression_elimination() {
+            let parsed_rule_from = parse_rule("#a * _b + #a * _b");
+            let parsed_rule_to = parse_rule("#a * _b + #a * _b");
+            let parsed_target = parse_expr("0 * 0 + 0 * 0");
+
```

Comment:
why are parsed_rule_from and parsed_rule_to the same thing. Just set one variable then. If this is just for the exercise, add a comment for the future.

Reply:
Good spot, I'll make them the same var